### PR TITLE
Sanitize Mmds request URI

### PR DESCRIPTION
**Pull request by sandreim**
## Reason for This PR

Fixes: #1963 

## Description of Changes

Provided in commit messages.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: src/mmds/src/lib.rs [hunk 1]
```diff
@@ -65,6 +65,20 @@ fn build_response(http_version: Version, status_code: StatusCode, body: Body) ->
...
 
+// Make the URI a correct JSON pointer value.
+fn sanitize_uri(uri: &str) -> String {
+    let mut sanitized_uri = uri.to_owned();
+    let mut len = u32::MAX as usize;
+    // Loop while the deduping decreases the sanitized len.
+    // Each iteration will attempt to dedup "//".
+    while sanitized_uri.len() < len {
+        len = sanitized_uri.len();
+        sanitized_uri = sanitized_uri.replace("//", "/");
```
Comment (iulianbarbu) [inline src/mmds/src/lib.rs:76]: Does it make sense to use a pattern and replace all the occurrences of "//+" with "/"? This wouldn't involve a while loop to reduce the "//" resulted from previous reduction in steps.

File: src/mmds/src/lib.rs [hunk 2]
```diff
@@ -65,6 +65,20 @@ fn build_response(http_version: Version, status_code: StatusCode, body: Body) ->
     response
 }
 
+// Make the URI a correct JSON pointer value.
+fn sanitize_uri(uri: &str) -> String {
+    let mut sanitized_uri = uri.to_owned();
+    let mut len = u32::MAX as usize;
+    // Loop while the deduping decreases the sanitized len.
```
Comment (lauralt) [inline src/mmds/src/lib.rs:72]:
```suggestion
    // Loop until the sanitized len is no longer modified in an iteration. 
```

File: src/mmds/src/lib.rs (hunk 1 above)
Comment (lauralt) [reply src/mmds/src/lib.rs:76]: I was searching that too but it seems rust doesn't offer support for regex in the std library.

Comment (sandreim) [reply src/mmds/src/lib.rs:76]: That sounds like we need to pull in some extra dependency to use regular expressions. This code is not on any hotpath so I do not want to make such trade-offs,

File: src/mmds/src/lib.rs [hunk 3]
```diff
@@ -65,6 +65,20 @@ fn build_response(http_version: Version, status_code: StatusCode, body: Body) ->
     response
 }
 
+// Make the URI a correct JSON pointer value.
+fn sanitize_uri(uri: &str) -> String {
```
Comment (lauralt) [inline src/mmds/src/lib.rs:69]: Would it make sense to make `uri` a String here and avoid the following line?

File: src/mmds/src/lib.rs (hunk 2 above)
Comment (sandreim) [reply src/mmds/src/lib.rs:72]: I think my initial comment has same meaning, why do you prefer it to be longer ?

File: src/mmds/src/lib.rs (hunk 3 above)
Comment (sandreim) [reply src/mmds/src/lib.rs:69]: The `uri` variable is a &str as we retrieve it here. I would still have to call to_string() somewhere else.
```
let uri = request.uri().get_abs_path();
```

File: src/mmds/src/lib.rs (hunk 2 above)
Comment (lauralt) [reply src/mmds/src/lib.rs:72]: For me the comment wasn't clear about when does the looping end and it seemed that this was its purpose. From the comment I'm thinking of an infinite loop. Anyway, it was more of a nit :D.

File: src/mmds/src/lib.rs (hunk 3 above)
Comment (lauralt) [reply src/mmds/src/lib.rs:69]: I know, the idea here was to call to_string() when calling `sanitize_uri()`, not inside of it; it was a matter of having one less code line.

Comment (sandreim) [reply src/mmds/src/lib.rs]: Done.

File: src/mmds/src/lib.rs (hunk 1 above)
Comment (iulianbarbu) [reply src/mmds/src/lib.rs]: In this case, I am OK with the while loop.

**Pull request merged**